pkg/release/helm/helm.go

```diff
@@ -0,0 +1,205 @@
+package helm
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"strings"
+
+	"github.com/google/go-github/v35/github"
+	"github.com/pkg/errors"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/utils/pointer"
+
+	"github.com/cert-manager/release/pkg/release/manifests"
+)
+
+const (
+	// Some template strings for GitHub commit and PR messages.
+	tplCommitMessage = "Add %s"
+	tplPRTitle       = "Helm charts for cert-manager staged release: %s"
+	tplPRDescription = "This PR was generated by [cmrel](https://github.com/cert-manager/release) for the cert-manager staged release: `%s`."
+)
+
+var acceptableGitHubPermissions sets.String
+
+func init() {
+	acceptableGitHubPermissions = sets.NewString("write", "admin")
+}
+
+type RepositoryManager interface {
+	// Check is called immediately after instantiating the RepositoryManager to
+	// verify that the supplied GitHub configuration is valid and that the
+	// supplied GitHub client has permissions to perform branching, uploading
+	// and creating PRs.
+	Check(ctx context.Context) error
+	// Publish creates a new GitHub branch, uploads all the chart files and then
+	// creates a PR to merge those into the main branch for the repository.
+	// The PR URl is returned on success.
+	Publish(ctx context.Context, releaseName string, charts ...manifests.Chart) (prURL string, err error)
+}
+
+type gitHubRepositoryManager struct {
+	*GitHubClient
+	owner  string
+	repo   string
+	branch string
+}
+
+// NewGitHubRepositoryManager returns a gitHubRepositoryManager which implements
+// RepositoryManager to upload Helm charts to a branch in a GitHub repository
+// and create a PR.
+func NewGitHubRepositoryManager(client *GitHubClient, owner, repo, branch string) RepositoryManager {
+	return &gitHubRepositoryManager{
+		GitHubClient: client,
+		owner:        owner,
+		repo:         repo,
+		branch:       branch,
+	}
+}
+
+// Check is documented at RepositoryManager.Check
+// Requests the user associated with the current Oauth token and checks whether
+// that user has write permission to the configured repository and whether it
+// has the repo scope.
+func (o *gitHubRepositoryManager) Check(ctx context.Context) (err error) {
+	var errs []error
+	defer func() {
+		err = errors.WithStack(utilerrors.NewAggregate(errs))
+	}()
```

Comment:
I thought it was useful to return an agregate error here, with details of *all* the failed checks.
To avoid the frustration of having to run `cmrel publish` multiple times in order to discover and satisfy all the check requirements.

The `WithStack` wrapper is useful in the integration test, because it shows the stacktrace if there's an error and allows you to easily see where the error originated.

Reply:
Think I mentioned it somewhere already- really nice error checking in this PR 👍🏼